design/apidsl/security.go

```diff
@@ -0,0 +1,308 @@
+// "basic".
+//
+// Example:
+//
+//     OtherSecurity("jwt", "apiKey", func() {
+//	    	InHeader("Authorization")
+//	    	AuthorizationURL("http://example.com/authorization")
+//	    	Scope("my_system:write", "Write to the system")
+//	    	Scope("my_system:read", "Read anything in there")
+//     })
+//
+func OtherSecurity(name string, securityType string, dsl ...func()) {
+	if _, ok := dslengine.CurrentDefinition().(*design.APIDefinition); !ok {
+		dslengine.IncompatibleDSL()
+		return
+	}
+
+	if securityMethodRedefined(name) {
+		return
+	}
+
+	def := &design.SecurityMethodDefinition{
+		Kind:   design.OtherSecurityKind,
+		Method: name,
+		Type:   securityType,
+	}
+
+	if len(dsl) != 0 {
+		if !dslengine.Execute(dsl[0], def) {
+			return
+		}
+	}
+
+	design.Design.SecurityMethods = append(design.Design.SecurityMethods, def)
+}
+
+// Scope defines an authorization scope. Used within SecurityMethod,
+// the description is required, explaining what the scope
+// means. Within a Security block, only a scope is needed.
+func Scope(name string, desc ...string) {
+	switch parent := dslengine.CurrentDefinition().(type) {
+	case *design.SecurityDefinition:
+		if len(desc) == 1 {
+			dslengine.ReportError("must not pass description to Scope()")
+			return
+		}
+		parent.Scopes = append(parent.Scopes, name)
+	case *design.SecurityMethodDefinition:
+		if len(desc) == 0 {
+			dslengine.ReportError("no description provided for Scope()")
+			return
+		}
+		if parent.Scopes == nil {
+			parent.Scopes = make(map[string]string)
+		}
+		parent.Scopes[name] = desc[0]
+	default:
+		dslengine.IncompatibleDSL()
+	}
+}
+
+// InHeader defines both `in` and `name` of a SecurityMethod.  See
+// http://swagger.io/specification/#securitySchemeObject for details.
+func InHeader(headerName string) {
+	if parent, ok := dslengine.CurrentDefinition().(*design.SecurityMethodDefinition); ok {
+		if parent.Kind == design.APIKeySecurityKind || parent.Kind == design.OtherSecurityKind {
+			if parent.In != "" {
+				dslengine.ReportError("'In' previously defined through InHeader or InQuery")
+				return
+			}
+			parent.In = "header"
+			parent.Name = headerName
+			return
+		}
+	}
+	dslengine.IncompatibleDSL()
+}
+
+// InQuery defines both `in` and `name` of a SecurityMethod. See
+// http://swagger.io/specification/#securitySchemeObject for details.
+func InQuery(parameterName string) {
+	if parent, ok := dslengine.CurrentDefinition().(*design.SecurityMethodDefinition); ok {
+		if parent.Kind == design.APIKeySecurityKind || parent.Kind == design.OtherSecurityKind {
+			if parent.In != "" {
+				dslengine.ReportError("'In' previously defined through InHeader or InQuery")
+				return
+			}
+			parent.In = "query"
+			parent.Name = parameterName
+			return
+		}
+	}
+	dslengine.IncompatibleDSL()
+}
+
+// OAuth2Flow defines the `flow` for a SecurityMethod.  See
+// http://swagger.io/specification/#securitySchemeObject for details.
```

Comment:
Yeah, tough for me to tell as I don't use OAuth2.

I understand your comment about Swagger, and us wanting to go above the spec and I agree with that. Even with the JWT case I have at hand, which needs to use the `OtherSecurity` method to work - I'd probably prefer to have a JWTSecurity and model all the scope thing on top - but the main reason I don't just do my own middleware and stick it in with `service.Use` directly, is that I want the swagger UI and swagger tooling.. so in a sense, being very close to the spec is desirable, especially for the cases that use its constraints, in order to be able to generate a meaningful and fully working Swagger UI out of the DSL.

Now this doesn't prevent from using a DSL for the Flow values, or describing more of what we'd find in the spec directly in our comments. Both of those make sense.

I don't see much value in drifting from the precise layout of swagger either, unless we have use cases.

OAuth2Flow could be just Flow.. this is where I took the liberty to namespace according to its semantic: it makes sense only in OAuth2. Not attached to it though :-)

I see that having some more refined Flow functions could make the DSL more robust, and discoverable.. I'm not strongly opinionated about it since I don't currently use it :-)

Regarding InQuery and InHeader.. I think itnwould make sense. Isn't there a risk that Header's semantic becomes unclear ? Sometimes, I find this leads to confusion, not knowing the effect of certain DSL funcs in certain contexts, especially when some are aliases of others. I remember being confused when I read that a Member was an alias to Attribute.

(I'm on mobile and the UI refreshes when I submit a note and I lose track of where I was so I did all comments in here)